Update dockerfile

terencehonles opened the pull request:
This builds off of #6 (and we can still work off of #6, but just opening this up as a suggestion)

- It flattens all the system layers into one, and then leaves the pyenv install as it's own layer (so the base image can be cached separately from the pyenv install)
- The build is much more verbose (to more easily tell when and where something went wrong)
- I have added the mandatory `--build-arg PYENV_UPDATED_AT=$(date +%Y-%m-%d)` and the build will fail saying that exact argument should be added.
- Installs only the most recent point release versions of Python
- Accepts an optional `--build-arg PYENV_EXTRA_PYTHON_VERSIONS='3.4.2 3.6.8'` to add explicit point releases which aren't the most recent point version.

I personally use `DOCKER_BUILDKIT=1` when building my images, and it may make sense to leverage [its caching](https://github.com/moby/buildkit/blob/dockerfile/1.1.7-experimental/frontend/dockerfile/docs/experimental.md#run---mounttypecache) to not have to build python versions over and over again. This would require using its header `# syntax = docker/dockerfile:1.1.7-experimental` and builds not using the `DOCKER_BUILDKIT=1` variable would fail.

If you were building via a CI (github action) I'm not sure how easily/long you can cache the buildkit cache so it might not actually make a difference, but it's something we use on our CI (we own it so we control the caching) to save on repeat downloads for pip/npm.

randomknowledge committed:
Update versions
* Ubuntu 16 -> 20
* node 8 -> 12
* remove python 2
* remove fabric

randomknowledge committed:
remove python version 3.0.0, because it doesn't exist for pyenv

randomknowledge committed:
use ubuntu 18.04, because ubuntu 20.04 doesn't support libssl1.0 which is needed by older python versions

randomknowledge committed:
add LC_ALL and PYTHONIOENCODING

terencehonles committed:
Update dockerfile to have less layers and use pyenv to determine versions

The versions installed automatically are the most recent point releases
for each of the Python versions starting at Python 2.7.

randomknowledge commented:
Hi @terencehonles,
sorry for the late reply, I'm quite busy atm.
Could you define a default value of `"3.0.1 3.1.5 3.2.6 3.3.7 3.4.2 3.4.10 3.5.7 3.6.8 3.7.3"` `for PYENV_EXTRA_PYTHON_VERSIONS`?

terencehonles committed:
Add GitHub workflow to build docker image

terencehonles commented:
> Hi @terencehonles,
> sorry for the late reply, I'm quite busy atm.
> Could you define a default value of `"3.0.1 3.1.5 3.2.6 3.3.7 3.4.2 3.4.10 3.5.7 3.6.8 3.7.3"` `for PYENV_EXTRA_PYTHON_VERSIONS`?

Just to call something out on that request:

Versions `3.0.1 3.1.5 3.2.6 3.3.7 3.4.10` will be included automatically since they are the most recent point releases. I added the versions `3.4.2 3.5.7 3.6.8 3.7.3` as part of the github action I added with the PR, but I did not include them in the Dockerfile itself. Some of those versions did not exist in the original docker image. Did you need them or should the list you provided be restricted to versions that were previously built in the existing docker image?

Regarding the GitHub workflow you will need to set up two GitHub secrets:
- DOCKERHUB_USERNAME
- DOCKERHUB_TOKEN

You can read more about that here: https://github.com/marketplace/actions/docker-login#dockerhub / https://docs.docker.com/docker-hub/access-tokens/

terencehonles commented:
Just a note, you'll want to disable docker building the Dockerfile from Docker Hub assuming you like the idea of the github action. I didn't provide the ability to rebuild the complete image, and just refreshed pyenv, but changes to the base image will trigger a complete rebuild so that might be sufficient to pull in any security updates.
